File: algorithms/JavaScript/src/linked-lists/singly.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class SinglyLinkedList {
  constructor() {
    this.head = null;
  }

  // Will return the element at that particular index in linklist starting
  // from head.
  getByIndex(idx) {
    if (!this.head) {
      return 'List is Empty';
    }

    let currentNode = this.head;
    let iterationCount = 0;

    for (
      currentNode = this.head;
      currentNode != null;
      currentNode = currentNode.next
    ) {
      if (iterationCount == idx) {
        return currentNode;
      }
      iterationCount++;
    }

    return null;
  }

  setByIndex(idx, newValue) {
    const nodeElem = this.getByIndex(idx);
    if (nodeElem) {
      nodeElem.data = newValue;
      return nodeElem;
    }

    return null;
  }
  insertAtHead(data) {
    const newNode = new Node(data);
    if (this.head == null) {
      this.head = newNode;
    } else {
      newNode.next = this.head;
      this.head = newNode;
    }
  }

  // idx : index of the node after which you want to
  // insert the new node;
  // data: value associated with new node.
  insertAfterByIndex(idx, data) {
    let afterNode;
    let newNode;

    const beforeNode = this.getByIndex(idx);

    if (beforeNode) {
      afterNode = beforeNode.next;
      newNode = new Node(data);

      beforeNode.next = newNode;
      newNode.next = afterNode;

      return this;
    }

    return null;
  }

  // idx: index of the node which you want to remove from linklist.
  removeFromMiddle(idx) {
    const beforeNode = this.getByIndex(idx - 1);
    let afterNode;
    let nodeToRemove;

    if (beforeNode) {
      nodeToRemove = beforeNode.next;

      if (nodeToRemove) {
        afterNode = nodeToRemove.next;
        beforeNode.next = afterNode;

        return this;
      }
    }
    return null;
  }

  removeAtHead() {
    // check for underflow
    if (this.head == null) {
      console.log('Underflow');
      return;
    }
    // store head's reference data value in local data variable
    const data = this.head.data;
    this.head = this.head.next;
    return data;
  }

  insertAtEnd(data) {
    const newNode = new Node(data);
    if (!this.head) {
      this.head = newNode;
      return this;
    }

    let currentNode = this.head;

    while (currentNode.next) {
      currentNode = currentNode.next;
    }

    currentNode.next = newNode;

    return this;
  }

  // Removes elements which are at the end of the linklist.
  removeFromEnd() {
    if (!this.head) {
      console.log('List is Empty');
      return this;
    }

    let returnNode;
    if (!this.head.next) {
      returnNode = this.head;
      this.head = null;
      return returnNode;
    }

    let currentNode = this.head.next;
    let previousNode = this.head;

    while (currentNode) {
      if (currentNode.next) {
        previousNode = currentNode;
      }
      currentNode = currentNode.next;
    }
    returnNode = previousNode.next;
    previousNode.next = null;
    return returnNode;
  }

  rotate() {
    if (!this.head) {
      return;
    }

    let currentNode = this.head;
    let previousNode = null;
    let tempNextRefNode;

    while (currentNode) {
      tempNextRefNode = currentNode.next;
      currentNode.next = previousNode;

      previousNode = currentNode;
      currentNode = tempNextRefNode;
    }

    this.head = previousNode;

    return this;
  }

  printList() {
    // if head is null then list is empty
    if (this.head == null) {
      console.log('List is Empty');
    } else {
      // iterate through list until itr is not null
      let itr = this.head;
      while (itr != null) {
        console.log(itr.data);
        itr = itr.next;
      }
    }
  }
}

const array = new SinglyLinkedList();
array.insertAtHead(1);
array.insertAtHead('xyz');
array.insertAtHead(1.1);
array.removeAtHead();
array.insertAtEnd(99);
array.insertAtEnd(100);
array.insertAtEnd(101);
array.insertAtEnd(102);
array.insertAtEnd(103);
array.insertAtEnd(104);
array.insertAtEnd(105);

array.printList();

array.removeFromEnd();

array.printList();

array.rotate();

array.printList();

File: algorithms/CPlusPlus/Linked-Lists/singly.cpp
#include <iostream>
#include <exception>

using dType = int;


template <typename T>
class SinglyLinkedList {
    private:
        template <typename N>
        class Node {
            public:
                T data;
                Node *next;

                Node(T data)
                    : data(data) , next(nullptr) {};
                Node(){}
        };

        Node<T>* head;
        Node<T>* tail;

    public:
        SinglyLinkedList()
            : head(nullptr),tail(nullptr){}

        ~SinglyLinkedList(){
            Node<T> *itr = this->head;
            // iterate through list until itr reach to nullptr and print data
            while(itr != nullptr) {
                // creating temporary pointer to delete memory
                Node<T> *temp_ptr = itr;
                itr = itr->next;
                delete temp_ptr;
            }

            // std::cout << "Destructor called\n";
        }

        void insertAtHead(T data) {
            Node<T> *temp = new Node<T>(data);

            // if head is not pointing to anything then simply point to new node
            if(this->head == nullptr) {
                this->head = temp;
                return;
            }

            // if head is already point at some node
            // first make new node points to head address
            temp->next = this->head;
            // change head point to new node
            this->head = temp;
        }

        void insertAtEnd(T data){

            Node<T> *newNode = new Node<T>(data);

            // if head is not pointing to anything then simply point to new node
            if(this->head == nullptr) {
                this->head = newNode;
                this->tail = newNode;
                return;
            }

            // if list is not empty, then put the address of the new node in the tail->next
            // and finally, make the next of newNode = NULL and tail will be pointing the the newNode as now it is in the last position.
            tail->next = newNode;
            newNode->next= NULL;
            tail=newNode;
        }

        T removeAtHead() {
            // if list is empty we can't remove node
            if(this->head == nullptr)
                throw std::runtime_error("List is Empty.");

            // save data of node where head is pointed
            T data = this->head->data;
            // save temp pointer to free the memory
            Node<T>* temp_ptr = this->head;
            this->head = this->head->next;

            // free memory
            delete temp_ptr;
            return data;
        }

        T removeAtEnd() {
            // if list is empty we can't remove node
            if (this->head == nullptr)
                throw std::runtime_error("List is Empty.");
            // if list is singleton it just becomes nullptr
            if(this->head->next == nullptr) {
                T data = this->tail->data;

                Node<T>* temp_ptr = this->tail;
                this->head = nullptr;
                this->tail = nullptr;

                delete temp_ptr;
                return data;
            }

            // save data of node where tail is pointed
            T data = this->tail->data;

            // now iterate to one before last pointer
            Node<T>* currNode = this->head;
            Node<T>* nextNode = currNode->next;

            while (nextNode && nextNode->next)
            {
                currNode = nextNode;
                nextNode = nextNode->next;
            }

            // save temp pointer to free the memory
            Node<T>* temp_ptr = this->tail;
            currNode->next = nullptr;
            this->tail = currNode;

            delete temp_ptr;
            return data;
        }

        T front() {
            if(this->head == nullptr)
                throw std::runtime_error("Empty List.");
            return this->head->data;
        }

        T back() {
            if(this->head == nullptr)
                throw std::runtime_error("Empty List.");
            return this->tail->data;
        }

        void printList() const {
            if(this->head == nullptr) {
                std::cout<<"List is Empty."<<std::endl;
            } else {
                Node<T> *itr = this->head;
                // iterate through list until itr reach to nullptr and print data
                while(itr != nullptr) {
                    std::cout<<itr->data<<" ";
                    itr = itr->next;
                }
                std::cout<<std::endl;
            }
        }
};

template <typename T>
void insert_element(SinglyLinkedList<T>& list, bool front) {
    T val {};
    std::cout << "Enter value to be inserted: ";
    std::cin >> val;

    if (front) list.insertAtHead(val);
    else list.insertAtEnd(val);
}

void menu() {
    SinglyLinkedList<dType> list {};
    while (true) {
        std::cout << "\nWhat would you like to do?\n";
        std::cout << "1. Push back an element\n";
        std::cout << "2. Push front an element\n";
        std::cout << "3. Display the linked list\n";
        std::cout << "4. Get front\n";
        std::cout << "5. Get back\n";
        std::cout << "6. Pop front\n";
        std::cout << "7. Pop back\n";
        std::cout << "8. Quit\n";
        std::cout << "Enter corresponding number to your choice: ";
        int ch {};
        std::cin >> ch;

        try {
            switch (ch) {
                case 1: insert_element(list, false);
                          break;
                case 2: insert_element(list, true);
                          break;
                case 3: list.printList();
                          break;
                case 4: std::cout << list.front() << "\n";
                          break;
                case 5: std::cout << list.back() << "\n";
                          break;
                case 6: list.removeAtHead();
                          break;
                case 7: list.removeAtEnd();
                          break;
                case 8: return;
                default: std::cout << "That's an invalid option. Please try again.\n";
            }
        }
        catch (const std::runtime_error& re) {
            std::cerr << re.what() << "\n";
        }
    }
}

int main() {
    menu();
}
